### PR TITLE
add data structures for entries in nodl xml files

### PR DESCRIPTION
**Pull request by Arnatious**
Basic data structures parser will populate based on https://github.com/ros2/design/blob/ef5018b583659f44bea503763d30b0c19a3818ef/articles/ros_node_idl.md

Signed-off-by: Ted Kern <ted.kern@canonical.com>

### CODE REVIEW

File: nodl_python/nodl/nodl_types.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Limited General Public License for more details.
+#
+# You should have received a copy of the GNU Limited General Public License along with this program.
+# If not, see <http://www.gnu.org/licenses/>.
+
+
+from rclpy.qos import QoSProfile, QoSPresetProfiles
+
+
+class nodl_action():
```
Comment (kyrofa) [inline nodl_python/nodl/nodl_types.py]: Can we capitalize and use pascal case for class names, please? Also, the parens are unnecessary.

File: nodl_python/nodl/nodl_types.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    """Data structure for action entries in NoDL."""
+
+    def __init__(self,
+                 name: str,
+                 msg_type: str,
```
Comment (kyrofa) [inline nodl_python/nodl/nodl_types.py]:
```suggestion
                 action_type: str,
```

File: nodl_python/nodl/nodl_types.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                 server: bool = False,
+                 client: bool = False,
+                 qos: QoSProfile = QoSPresetProfiles.ACTION_STATUS_DEFAULT):
+        self.name = name
+        self.type = msg_type
```
Comment (kyrofa) [inline nodl_python/nodl/nodl_types.py]:
```suggestion
        self.type = action_type
```

File: nodl_python/nodl/nodl_types.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+
+from rclpy.qos import QoSProfile, QoSPresetProfiles
+
+
+class nodl_action():
+    """Data structure for action entries in NoDL."""
+
+    def __init__(self,
+                 name: str,
```
Comment (kyrofa) [inline nodl_python/nodl/nodl_types.py]: Given that this is an API, we should probably force keywords-only.

File: nodl_python/nodl/nodl_types.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        self.qos = qos
+
+
+class nodl_service():
+    """Data structure for service entries in NoDL."""
+
+    def __init__(self,
+                 name: str,
+                 msg_type: str,
```
Comment (kyrofa) [inline nodl_python/nodl/nodl_types.py]:
```suggestion
                 service_type: str,
```

File: nodl_python/nodl/nodl_types.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                 server: bool = False,
+                 client: bool = False,
+                 qos: QoSProfile = QoSPresetProfiles.SERVICES_DEFAULT):
+        self.name = name
+        self.type = msg_type
```
Comment (kyrofa) [inline nodl_python/nodl/nodl_types.py]:
```suggestion
        self.type = service_type
```

File: nodl_python/nodl/nodl_types.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+        self.qos = qos
+
+
+class nodl_topic():
+    """Data structure for topic entries in NoDL."""
+
+    def __init__(self,
+                 name: str,
+                 msg_type: str,
```
Comment (kyrofa) [inline nodl_python/nodl/nodl_types.py]: Too verbose to use "message" here, you think?

File: nodl_python/nodl/nodl_types.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                 server: bool = False,
+                 client: bool = False,
+                 qos: QoSProfile = QoSPresetProfiles.ACTION_STATUS_DEFAULT):
+        self.name = name
+        self.type = msg_type
+        self.server = server
+        self.client = client
```
Comment (artivis) [inline nodl_python/nodl/nodl_types.py]: Might be worth it to do some sanity check here, 
```python
assert self.server != self.client, "Either a server or a client, not both, not none!"
```

File: nodl_python/nodl/nodl_types.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    def __init__(self,
+                 name: str,
+                 msg_type: str,
+                 server: bool = False,
+                 client: bool = False,
+                 qos: QoSProfile = QoSPresetProfiles.SERVICES_DEFAULT):
+        self.name = name
+        self.type = msg_type
+        self.server = server
+        self.client = client
```
Comment (artivis) [inline nodl_python/nodl/nodl_types.py]: See above sanity check comment.

File: nodl_python/nodl/nodl_types.py (hunk 8 above)
Comment (artivis) [reply nodl_python/nodl/nodl_types.py]: Actually, maybe server and client aren't mutually exclusive, `assert self.server or self.client`.

File: nodl_python/nodl/nodl_types.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        self.qos = qos
+
+
+class nodl_topic():
+    """Data structure for topic entries in NoDL."""
+
+    def __init__(self,
+                 name: str,
+                 msg_type: str,
+                 server: bool = False,
```
Comment (artivis) [inline nodl_python/nodl/nodl_types.py]:
```suggestion
                 publisher: bool = False,
```

File: nodl_python/nodl/nodl_types.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+                 client: bool = False,
```
Comment (artivis) [inline nodl_python/nodl/nodl_types.py]:
```suggestion
                 subscription: bool = False,
```

File: nodl_python/nodl/nodl_types.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                 qos: str = QoSPresetProfiles.SYSTEM_DEFAULT):
+        self.name = name
+        self.type = msg_type
+        self.server = server
```
Comment (artivis) [inline nodl_python/nodl/nodl_types.py]:
```suggestion
        self.publisher = publisher
```

File: nodl_python/nodl/nodl_types.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        self.client = client
```
Comment (artivis) [inline nodl_python/nodl/nodl_types.py]:
```suggestion
        self.subscription = subscription
```

Comment (artivis) [inline nodl_python/nodl/nodl_types.py]: See above sanity check comment.

File: nodl_python/nodl/nodl_types.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without even the implied warranties of MERCHANTABILITY, SATISFACTORY QUALITY,
+# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Limited General Public License for more details.
+#
+# You should have received a copy of the GNU Limited General Public License along with this program.
+# If not, see <http://www.gnu.org/licenses/>.
+
+
+from rclpy.qos import QoSProfile, QoSPresetProfiles
+
```
Comment (artivis) [inline nodl_python/nodl/nodl_types.py]: It seems that a `nodl_parameter` class is missing.

Comment (artivis) [reply nodl_python/nodl/nodl_types.py]: Resolved in fd8a77ad2347dacad2eba422d3e6cf442c648e57.

File: nodl_python/nodl/nodl_types.py (hunk 8 above)
Comment (artivis) [reply nodl_python/nodl/nodl_types.py]: Actually, maybe having both set to `false` should trigger a warning rather than an assert :D

Comment (Arnatious) [reply nodl_python/nodl/nodl_types.py]: That should be done in the parser, these datatypes should be passive, essentially just mypy type checked dicts

Comment (kyrofa) [reply nodl_python/nodl/nodl_types.py]: You're right @artivis, that will definitely need to happen, but I agree with @Arnatious that this is a low enough level it's not necessary. These are essentially just PODs, the logic they require will be a level up.

**Comment by kyrofa** [comment]
I went ahead and fixed the flake8 issues.

**Pull request merged**